### PR TITLE
Harden chunked log probabilities

### PR DESCRIPTION
**Pull request by kashif**
# What does this PR do?

Adds LM-head bias support and re-gathers partitioned weights during `_ChunkedLogProbFunction.backward`. Includes full-logit loss and gradient parity tests. Part of #7063.

## Before submitting

- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request), Pull Request section?
- [x] Was this discussed/approved via a GitHub issue? Please add a link to it if that's the case.
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

## AI writing disclosure

We welcome the use of AI tools to help with contributions. For transparency and to help us improve our review process, please indicate the level of AI involvement in this PR.

- [ ] No AI usage: the PR was written entirely by a human.
- [ ] AI-assisted: some parts were suggested or improved by AI, but the PR was written and reviewed by a human.
- [x] AI-generated: the PR was mostly or fully generated by an AI tool.

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag members/contributors who may be interested in your PR.

---

> [!NOTE]
> **Medium Risk**
> Changes core RL/training loss math and gradients for chunked LM-head log-probs, including ZeRO-3 backward gathering; incorrect bias or gather behavior would skew policy gradients, though tests cover bias parity and existing backward paths.
> 
> **Overview**
> Extends the memory-efficient **`_ChunkedLogProbFunction`** path so chunked per-token log-probs and entropy match models whose **`lm_head` has bias**, and so backward stays correct under **DeepSpeed ZeRO-3** when weights are read outside a normal `lm_head` forward.
> 
> **`_ChunkedLogProbFunction`** now takes an optional **`bias`** argument (inserted after `weight`). Forward adds the bias slice per vocab chunk; backward accumulates **`grad_bias`** and wraps the recomputation loop in **`maybe_gather_lm_head_ctx(weight, bias)`** so partitioned head parameters are gathered during backward like fused losses need on forward. **`patch_chunked_lm_head`** passes **`self.lm_head.bias`** into `apply`.
> 
> Tests pass **`None`** for the new bias slot on existing cases and add **`test_bias`** (float32 and bfloat16) comparing forward, combined loss, and gradients to a reference that includes bias. **`maybe_gather_lm_head_ctx`** docs are updated to note that custom autograd functions must re-enter the context on backward.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit fd259e0c46018a8d941dc563a7174703ff8e9373. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Review by qgallouedec — APPROVED** [review]
good!

File: trl/trainer/utils.py [hunk 1]
```diff
@@ -1395,71 +1398,86 @@ def forward(
...
-            grad_weight[start:end].add_(grad_logits.t() @ hidden.float())
-
-        return grad_hidden.to(hidden.dtype), grad_weight.to(weight.dtype), None, None, None, None, None
+        with maybe_gather_lm_head_ctx(weight, bias):
+            vocab = weight.shape[0]
+
+            # NOTE(@aminediro): always acc in fp32 even if input is not
+            grad_hidden = torch.zeros(hidden.shape, device=hidden.device, dtype=torch.float32)
+            grad_weight = torch.zeros(weight.shape, device=weight.device, dtype=torch.float32)
+            grad_bias = torch.zeros(bias.shape, device=bias.device, dtype=torch.float32) if bias is not None else None
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/trainer/utils.py:1415]: No, not the case: a leaf Parameter comes back from ctx.saved_tensors as the same Python object, ds_status and the other ds attributes are intact (checked with a Parameter carrying a custom attribute).

File: trl/trainer/utils.py [hunk 2]
```diff
@@ -1356,6 +1357,8 @@ def forward(
             # with last_hidden.dtype (float16) while w_chunk (the lm_head weights) is not auto casted
             w_chunk = weight[start:end].to(last_hidden.dtype)  # [C, H]
             torch.mm(last_hidden, w_chunk.t(), out=mm_buf[:, :C])
+            if bias is not None:
+                mm_buf[:, :C].add_(bias[start:end].to(last_hidden.dtype))
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/trainer/utils.py:1361]: DeepSpeed does this gather itself: attribute access to a partitioned param inside a module forward goes through `ZeROOrderedDict.__getitem__`, which all-gathers it and registers it as an external param, and the backward hooks re-fetch it

On 2 GPUs with `stage3_param_persistence_threshold: 0` (weight really at numel 0 outside the forward): forward and backward match the full-logits reference with no explicit gather, on this branch and on main. So no change needed here; the backward gather in this PR is belt-and-braces for callers outside a module forward.
